### PR TITLE
Fix Entity Operator status race condition

### PR DESCRIPTION
**Pull request by kyguy**
### Type of change

- Bugfix

### Description

Currently, there is a race condition between the Entity Operator Readiness check and Entity Operator replica set deletion, allowing the readiness check to pass on EO pods before they are deleted during an EO update. This can cause a Kafka Resource's Ready status to be marked as true, when the EO is failing.

This PR ensures that the EO readiness is not checked until the updated EO has been generated, preventing the readiness check to pass on a stale EO replica set.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/DeploymentOperator.java [hunk 1]
```diff
@@ -84,4 +84,24 @@ protected Integer currentScale(String namespace, String name) {
...
+    }
+
+    /**
+     * Check if a deployment has been generated.
+     *
+     * @param namespace The namespace.
+     * @param name The resource name.
+     * @return Whether the deployment has been generated.
+     */
+    public boolean isGenerated(String namespace, String name) {
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/DeploymentOperator.java]: I wonder if we can find some better name. `isGenerated` can have different expectations. What about `isObserved`?

Also, does this need to be public?

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/DeploymentOperator.java [hunk 2]
```diff
@@ -84,4 +84,24 @@ protected Integer currentScale(String namespace, String name) {
         Annotations.annotations(desired).put(Annotations.ANNO_DEP_KUBE_IO_REVISION, k8sRev);
         return super.internalPatch(namespace, name, current, desired, cascading);
     }
+
+    public Future<Void> generation(String namespace, String name, long pollIntervalMs, long timeoutMs) {
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/DeploymentOperator.java]: Add Javadoc please.

We should also here find some better name. We are trying to find out if the deployment has been updated. So maybe something based on that?

Comment (kyguy) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/DeploymentOperator.java]: Would `observing` be a good replacement for `generation`? Or would that be too general?

Comment (kyguy) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/DeploymentOperator.java]: I was of some other names like `updatedDeployment`, `latestDeployment`, `observingDeploymentGeneration`

**Comment by scholzj** [comment]
CC @stanlyDoge: Something like this will be needed for all the other statuses ... although maybe a bit more complicated because the rollback is different for Connect, MM and Bridge.

**Pull request merged**